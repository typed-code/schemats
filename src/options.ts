import { camelCase, upperFirst } from 'lodash';

const DEFAULT_OPTIONS: OptionValues = {
  writeHeader: true,
  camelCase: false,
};

export interface OptionValues {
  camelCase?: boolean;
  writeHeader?: boolean; // write schemats description header
}

export default class Options {
  public options: OptionValues;

  constructor(options: OptionValues = {}) {
    this.options = { ...DEFAULT_OPTIONS, ...options };
  }

  public transformTypeName(typename: string) {
    return this.options.camelCase ? upperFirst(camelCase(typename)) : typename;
  }

  public transformColumnName(columnName: string) {
    return this.options.camelCase ? camelCase(columnName) : columnName;
  }
}
