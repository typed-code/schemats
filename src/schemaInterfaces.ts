import Options from './options';

export interface ColumnDefinition {
  udtName: string;
  nullable: boolean;
  tsType?: string;
}

export interface TableDefinition {
  [columnName: string]: ColumnDefinition;
}

export interface Database {
  connectionString: string;

  query(queryString: string): Promise<object[]>;

  getDefaultSchema(): string;

  getEnumTypes(schema?: string): Promise<{ [key: string]: string[] }>;

  getTableDefinition(tableName: string, tableSchema: string): Promise<TableDefinition>;

  getTableTypes(tableName: string, tableSchema: string, options: Options): Promise<TableDefinition>;

  getSchemaTables(schemaName: string): Promise<string[]>;
}
