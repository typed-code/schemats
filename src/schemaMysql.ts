import { isEqual, keys, mapValues } from 'lodash';
import { Connection, createConnection, MysqlError } from 'mysql';
import { parse as urlParse } from 'url';
import { Options } from './options';
import { Database, ICustomTypes, ITable } from './schemaInterfaces';

export class MysqlDatabase implements Database {
  private db: Connection;
  private readonly defaultSchema: string;

  constructor(public connectionString: string) {
    this.db = createConnection(connectionString);
    const url = urlParse(connectionString, true);
    this.defaultSchema = url && url.pathname ? url.pathname.substr(1) : 'public';
  }

  // uses the type mappings from https://github.com/mysqljs/ where sensible
  private static mapTableDefinitionToType(
    tableDefinition: ITable,
    customTypes: string[],
    options: Options
  ): ITable {
    if (!options) {
      throw new Error('No options given');
    }

    tableDefinition.columns = mapValues(tableDefinition.columns, column => {
      switch (column.udtName) {
        case 'char':
        case 'varchar':
        case 'text':
        case 'tinytext':
        case 'mediumtext':
        case 'longtext':
        case 'time':
        case 'geometry':
        case 'set':
        case 'enum':
          // keep set and enum defaulted to string if custom type not mapped
          column.tsType = 'string';
          return column;
        case 'integer':
        case 'int':
        case 'smallint':
        case 'mediumint':
        case 'bigint':
        case 'double':
        case 'decimal':
        case 'numeric':
        case 'float':
        case 'year':
          column.tsType = 'number';
          return column;
        case 'tinyint':
          column.tsType = 'boolean';
          return column;
        case 'json':
          column.tsType = 'Object';
          return column;
        case 'date':
        case 'datetime':
        case 'timestamp':
          column.tsType = 'Date';
          return column;
        case 'tinyblob':
        case 'mediumblob':
        case 'longblob':
        case 'blob':
        case 'binary':
        case 'varbinary':
        case 'bit':
          column.tsType = 'Buffer';
          return column;
        default:
          if (customTypes.includes(column.udtName)) {
            column.tsType = options.transformTypeName(column.udtName);
            return column;
          } else {
            // tslint:disable-next-line
            console.log(
              `Type '${column.udtName}' has been mapped to 'any' because no specific type has been found.`
            );
            column.tsType = 'any';
            return column;
          }
      }
    });

    return tableDefinition;
  }

  private static parseMysqlEnumeration(mysqlEnum: string): string[] {
    return mysqlEnum
      .replace(/(^(enum|set)\('|'\)$)/gi, '')
      .split(`','`)
      .sort();
  }

  private static getEnumNameFromColumn(dataType: string, columnName: string): string {
    return `${columnName}_${dataType}`;
  }

  public query<T = object>(queryString: string) {
    return this.queryAsync<T>(queryString);
  }

  public async getEnumTypes(schema?: string, tables: string[] = []): Promise<ICustomTypes> {
    let additionWhereClause: string = '';
    const params: string[] = [];
    if (schema) {
      additionWhereClause = `and table_schema = ? `;
      params.push(schema);
    }

    if (tables.length > 0) {
      additionWhereClause = `and table_name in (${tables.map(_ => '?').join(',')}) `;
      params.push(...tables);
    }

    const rawEnumRecords = await this.queryAsync<{
      table_name: string;
      column_name: string;
      column_type: string;
      data_type: string;
    }>(
      'SELECT table_name, column_name, column_type, data_type ' +
        'FROM information_schema.columns ' +
        `WHERE data_type IN ('enum', 'set') ${additionWhereClause}ORDER BY table_name, column_name`,
      params
    );

    const groupedEnums = rawEnumRecords.reduce(
      (enums, { table_name, column_name, data_type, column_type }) => {
        const enumName = MysqlDatabase.getEnumNameFromColumn(data_type, column_name);
        const enumValues = MysqlDatabase.parseMysqlEnumeration(column_type);

        enums[enumName] = enums[enumName] || {};
        enums[enumName][table_name] = enumValues;

        return enums;
      },
      ({} as unknown) as { [enumName: string]: { [tableName: string]: string[] } }
    );

    return Object.entries(groupedEnums).reduce((result, [enumName, tablesMap]) => {
      const tableKeys = Object.keys(tablesMap);
      const firstTableValues = tablesMap[tableKeys[0]];
      const allValuesSame = tableKeys.every(tableKey =>
        isEqual(tablesMap[tableKey], firstTableValues)
      );

      if (tableKeys.length === 1 || allValuesSame) {
        result[enumName] = tablesMap[tableKeys[0]];
      } else {
        tableKeys.forEach(tableKey => {
          result[`${tableKey}_${enumName}`] = tablesMap[tableKey];
        });
      }

      return result;
    }, ({} as unknown) as ICustomTypes);
  }

  public async getTablesDefinition(
    tableNames: string[],
    tableSchema: string,
    customTypes: ICustomTypes
  ): Promise<ITable[]> {
    const params = [tableSchema];
    let whereClauseAddition = '';

    if (tableNames.length) {
      params.push(...tableNames);
      whereClauseAddition = `and table_name IN (${tableNames.map(_ => '?').join(',')}) `;
    }

    const tableColumns = await this.queryAsync<{
      table_name: string;
      column_name: string;
      data_type: string;
      is_nullable: 'YES' | 'NO';
    }>(
      'SELECT table_name, column_name, data_type, is_nullable ' +
        'FROM information_schema.columns ' +
        'WHERE table_schema = ? ' +
        whereClauseAddition +
        'ORDER BY table_name, column_name',
      params
    );

    const tablesMap = tableColumns.reduce(
      (result, { table_name, column_name, data_type, is_nullable }) => {
        result[table_name] = result[table_name] || {
          name: table_name,
          columns: {},
        };

        const table = result[table_name];

        const enumName = MysqlDatabase.getEnumNameFromColumn(data_type, column_name);

        table.columns[column_name] = {
          udtName: /^(enum|set)$/i.test(data_type)
            ? customTypes.hasOwnProperty(enumName)
              ? enumName
              : `${table_name}_${enumName}`
            : data_type,
          nullable: is_nullable === 'YES',
          tsType: '',
        };

        return result;
      },
      ({} as unknown) as { [tableName: string]: ITable }
    );

    return Object.values(tablesMap);
  }

  public async getTablesTypes(
    tableNames: string[],
    tableSchema: string,
    customTypes: ICustomTypes,
    options: Options
  ): Promise<ITable[]> {
    const customTypesKeys = keys(customTypes);

    const tableDefinitions = await this.getTablesDefinition(tableNames, tableSchema, customTypes);

    return tableDefinitions.map(table =>
      MysqlDatabase.mapTableDefinitionToType(table, customTypesKeys, options)
    );
  }

  public async getSchemaTables(schemaName: string): Promise<string[]> {
    const schemaTables = await this.queryAsync<{ table_name: string }>(
      'SELECT table_name ' +
        'FROM information_schema.columns ' +
        'WHERE table_schema = ? ' +
        'GROUP BY table_name ORDER BY table_name',
      [schemaName]
    );
    return schemaTables.map(({ table_name }) => table_name);
  }

  public getDefaultSchema(): string {
    return this.defaultSchema;
  }

  private queryAsync<T = object>(queryString: string, escapedValues?: string[]): Promise<T[]> {
    return new Promise((resolve, reject) => {
      this.db.query(queryString, escapedValues, (error: MysqlError, results: T[]) => {
        if (error) {
          return reject(error);
        }
        return resolve(this.toLowerCaseColumnName<T>(results));
      });
    });
  }

  private toLowerCaseColumnName<T>(results: T[]): T[] {
    return results.map((row: any) =>
      Object.keys(row).reduce((newRow, key) => {
        newRow[key.toLowerCase()] = row[key];
        return newRow;
      }, {} as any)
    );
  }
}
