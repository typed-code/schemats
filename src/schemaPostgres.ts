import { keys, mapValues } from 'lodash';
import * as PgPromise from 'pg-promise';
import { Options } from './options';

import { Database, ITable } from './schemaInterfaces';

const pgp = PgPromise();

export class PostgresDatabase implements Database {
  private db: PgPromise.IDatabase<{}>;

  constructor(public connectionString: string) {
    this.db = pgp(connectionString);
  }

  private static mapTableDefinitionToType(
    tableDefinition: ITable,
    customTypes: string[],
    options: Options
  ): ITable {
    tableDefinition.columns = mapValues(tableDefinition.columns, column => {
      switch (column.udtName) {
        case 'bpchar':
        case 'char':
        case 'varchar':
        case 'text':
        case 'citext':
        case 'uuid':
        case 'bytea':
        case 'inet':
        case 'time':
        case 'timetz':
        case 'interval':
        case 'name':
          column.tsType = 'string';
          return column;
        case 'int2':
        case 'int4':
        case 'int8':
        case 'float4':
        case 'float8':
        case 'numeric':
        case 'money':
        case 'oid':
          column.tsType = 'number';
          return column;
        case 'bool':
          column.tsType = 'boolean';
          return column;
        case 'json':
        case 'jsonb':
          column.tsType = 'Object';
          return column;
        case 'date':
        case 'timestamp':
        case 'timestamptz':
          column.tsType = 'Date';
          return column;
        case '_int2':
        case '_int4':
        case '_int8':
        case '_float4':
        case '_float8':
        case '_numeric':
        case '_money':
          column.tsType = 'Array<number>';
          return column;
        case '_bool':
          column.tsType = 'Array<boolean>';
          return column;
        case '_varchar':
        case '_text':
        case '_citext':
        case '_uuid':
        case '_bytea':
          column.tsType = 'Array<string>';
          return column;
        case '_json':
        case '_jsonb':
          column.tsType = 'Array<Object>';
          return column;
        case '_timestamptz':
          column.tsType = 'Array<Date>';
          return column;
        default:
          if (customTypes.indexOf(column.udtName) !== -1) {
            column.tsType = options.transformTypeName(column.udtName);
            return column;
          } else {
            // tslint:disable-next-line
            console.log(
              `Type '${column.udtName}' has been mapped to 'any' because no specific type has been found.`
            );
            column.tsType = 'any';
            return column;
          }
      }
    });
    return tableDefinition;
  }

  public query(queryString: string) {
    return this.db.query(queryString);
  }

  public async getEnumTypes(
    schema?: string,
    tables: string[] = []
  ): Promise<{ [key: string]: string[] }> {
    interface T {
      name: string;
      value: any;
    }

    const enums: any = {};
    const enumSchemaWhereClause = schema ? pgp.as.format(`where n.nspname = $1`, schema) : '';
    await this.db.each<T>(
      'select n.nspname as schema, t.typname as name, e.enumlabel as value ' +
        'from pg_type t ' +
        'join pg_enum e on t.oid = e.enumtypid ' +
        'join pg_catalog.pg_namespace n ON n.oid = t.typnamespace ' +
        `${enumSchemaWhereClause} ` +
        'order by t.typname asc, e.enumlabel asc;',
      [],
      (item: T) => {
        if (!enums[item.name]) {
          enums[item.name] = [];
        }
        enums[item.name].push(item.value);
      }
    );
    return enums;
  }

  public async getTablesDefinition(tableNames: string[], tableSchema: string): Promise<ITable[]> {
    interface T {
      table_name: string;
      column_name: string;
      udt_name: string;
      is_nullable: string;
    }

    const tablesMap: { [tableName: string]: ITable } = {};
    let whereClauseAddition = '';
    const params: any[] = [tableSchema];

    if (tableNames.length) {
      whereClauseAddition = `and table_name IN ($2:csv) `;
      params.push(tableNames);
    }

    await this.db.each<T>(
      'SELECT table_name, column_name, udt_name, is_nullable ' +
        'FROM information_schema.columns ' +
        `WHERE table_schema = $1 ${whereClauseAddition}` +
        'ORDER BY table_name, column_name',
      params,
      (schemaItem: T) => {
        tablesMap[schemaItem.table_name] = tablesMap[schemaItem.table_name] || {
          name: schemaItem.table_name,
          columns: {},
        };
        const table = tablesMap[schemaItem.table_name];

        table.columns[schemaItem.column_name] = {
          udtName: schemaItem.udt_name,
          nullable: schemaItem.is_nullable === 'YES',
          tsType: '',
        };
      }
    );

    return Object.values(tablesMap);
  }

  public async getTablesTypes(
    tableNames: string[],
    tableSchema: string,
    options: Options
  ): Promise<ITable[]> {
    const enumTypes = await this.getEnumTypes();
    const customTypes = keys(enumTypes);
    const tableDefinitions = await this.getTablesDefinition(tableNames, tableSchema);

    return tableDefinitions.map(table =>
      PostgresDatabase.mapTableDefinitionToType(table, customTypes, options)
    );
  }

  public async getSchemaTables(schemaName: string): Promise<string[]> {
    return this.db.map<string>(
      'SELECT table_name ' +
        'FROM information_schema.columns ' +
        'WHERE table_schema = $1 ' +
        'GROUP BY table_name ORDER BY table_name',
      [schemaName],
      (schemaItem: { table_name: string }) => schemaItem.table_name
    );
  }

  public getDefaultSchema(): string {
    return 'public';
  }
}
