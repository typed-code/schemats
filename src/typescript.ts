/**
 * Generate typescript interface from table schema
 * Created by xiamx on 2016-08-10.
 */

import * as _ from 'lodash';

import Options from './options';
import { TableDefinition } from './schemaInterfaces';

function nameIsReservedKeyword(name: string): boolean {
  const reservedKeywords = ['string', 'number', 'package'];
  return reservedKeywords.indexOf(name) !== -1;
}

function normalizeName(name: string, options: Options): string {
  return nameIsReservedKeyword(name) ? name + '_' : name;
}

export function generateTableInterface(
  tableNameRaw: string,
  tableDefinition: TableDefinition,
  options: Options
) {
  const tableName = options.transformTypeName(tableNameRaw);

  const members = Object.keys(tableDefinition)
    .map(c => options.transformColumnName(c))
    .reduce((result, columnName) => {
      result.push(`${columnName}: ${tableName}Fields.${normalizeName(columnName, options)};`);

      return result;
    }, [] as string[]);

  return `
  export interface ${normalizeName(tableName, options)} {
    ${members.join('\n    ')}
  }
`;
}

export function generateEnumType(enumObject: { [key: string]: string[] }, options: Options) {
  const enums = Object.entries(enumObject);
  return (
    enums
      .reduce((result, [enumNameRaw, values]) => {
        const enumName = options.transformTypeName(enumNameRaw);
        result.push(`export type ${enumName} = ${values.map(v => `'${v}'`).join(' | ')};`);

        return result;
      }, [] as string[])
      .join('\n') + (enums.length ? '\n' : '')
  );
}

export function generateTableTypes(
  tableNameRaw: string,
  tableDefinition: TableDefinition,
  options: Options
) {
  const tableName = options.transformTypeName(tableNameRaw);

  const fields = Object.entries(tableDefinition).reduce((result, [columnNameRaw, definition]) => {
    const type = definition.tsType;
    const nullable = definition.nullable ? ' | null' : '';
    const columnName = options.transformColumnName(columnNameRaw);
    result.push(`export type ${normalizeName(columnName, options)} = ${type}${nullable};`);
    return result;
  }, [] as string[]);

  return `
  export namespace ${tableName}Fields {
    ${fields.join('\n    ')}
  }
`;
}
