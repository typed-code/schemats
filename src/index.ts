/**
 * Schemats takes sql database schema and creates corresponding typescript definitions
 * Created by xiamx on 2016-08-10.
 */

import * as pkgUp from 'pkg-up';
import * as prettier from 'prettier';
import Options, { OptionValues } from './options';
import { Database, getDatabase } from './schema';
import { generateEnumType, generateTableInterface, generateTableTypes } from './typescript';

const pkgPath = pkgUp.sync();
const pkgVersion = pkgPath ? require(pkgPath).version : 'UNKNOWN';

function getTime() {
  const padTime = (value: number) => `0${value}`.slice(-2);
  const time = new Date();
  const yyyy = time.getFullYear();
  const MM = padTime(time.getMonth() + 1);
  const dd = padTime(time.getDate());
  const hh = padTime(time.getHours());
  const mm = padTime(time.getMinutes());
  const ss = padTime(time.getSeconds());
  return `${yyyy}-${MM}-${dd} ${hh}:${mm}:${ss}`;
}

function buildHeader(
  db: Database,
  tables: string[],
  schema: string | null,
  options: OptionValues
): string {
  const commands = [
    'schemats',
    'generate',
    '-c',
    db.connectionString.replace(/:\/\/.*@/, '://username:password@'),
  ];
  if (options.camelCase) {
    commands.push('-C');
  }
  if (tables.length > 0) {
    tables.forEach((t: string) => {
      commands.push('-t', t);
    });
  }
  if (schema) {
    commands.push('-s', schema);
  }

  return `
        /**
         * AUTO-GENERATED FILE @ ${getTime()} - DO NOT EDIT!
         *
         * This file was automatically generated by schemats v.${pkgVersion}
         * $ ${commands.join(' ')}
         *
         */

    `;
}

export async function typescriptOfTable(
  db: Database | string,
  table: string,
  schema: string,
  options = new Options()
) {
  if (typeof db === 'string') {
    db = getDatabase(db);
  }

  const tableTypes = await db.getTableTypes(table, schema, options);

  let interfaces = '';
  interfaces += generateTableTypes(table, tableTypes, options);
  interfaces += generateTableInterface(table, tableTypes, options);
  return interfaces;
}

export async function typescriptOfSchema(
  db: Database | string,
  tables: string[] = [],
  schema: string | null = null,
  options: OptionValues = {}
): Promise<string> {
  if (typeof db === 'string') {
    db = getDatabase(db);
  }

  if (!schema) {
    schema = db.getDefaultSchema();
  }

  if (tables.length === 0) {
    tables = await db.getSchemaTables(schema);
  }

  const optionsObject = new Options(options);

  const enumTypes = generateEnumType(await db.getEnumTypes(schema), optionsObject);
  const interfacePromises = tables.map(table =>
    typescriptOfTable(db, table, schema as string, optionsObject)
  );
  const interfaces = await Promise.all(interfacePromises).then(tsOfTable => tsOfTable.join(''));

  let output = '/* tslint:disable */\n\n';
  if (optionsObject.options.writeHeader) {
    output += buildHeader(db, tables, schema, options);
  }
  output += enumTypes;
  output += interfaces;

  const configFilePath = await prettier.resolveConfigFile();
  const prettierOptions = await prettier.resolveConfig(configFilePath || '');
  return prettier.format(output, { ...prettierOptions, parser: 'typescript' });
}

export { Database, getDatabase } from './schema';
export { Options, OptionValues };
