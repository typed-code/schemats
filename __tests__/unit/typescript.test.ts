import { Options } from '../../src/options';
import * as Typescript from '../../src/typescript';

const options = new Options({});

describe('Typescript', () => {
  describe('generateTableInterface', () => {
    it('empty table definition object', () => {
      const tableInterface = Typescript.generateTableInterface('tableName', {}, options);
      expect(tableInterface).toMatchSnapshot();
    });

    it('table name is reserved', () => {
      const tableInterface = Typescript.generateTableInterface('package', {}, options);
      expect(tableInterface).toMatchSnapshot();
    });

    it('table with columns', () => {
      const tableInterface = Typescript.generateTableInterface(
        'tableName',
        {
          col1: { udtName: 'name1', nullable: false },
          col2: { udtName: 'name2', nullable: false },
        },
        options
      );
      expect(tableInterface).toMatchSnapshot();
    });

    it('table with reserved columns', () => {
      const tableInterface = Typescript.generateTableInterface(
        'tableName',
        {
          string: { udtName: 'name1', nullable: false },
          number: { udtName: 'name2', nullable: false },
          package: { udtName: 'name3', nullable: false },
        },
        options
      );
      expect(tableInterface).toMatchSnapshot();
    });
  });

  describe('generateEnumType', () => {
    it('empty object', () => {
      const enumType = Typescript.generateEnumType({}, options);
      expect(enumType).toBe('');
    });

    it('with enumerations', () => {
      const enumType = Typescript.generateEnumType(
        {
          enum1: ['val1', 'val2', 'val3', 'val4'],
          enum2: ['val5', 'val6', 'val7', 'val8'],
        },
        options
      );
      expect(enumType).toBe(
        `export type enum1 = 'val1' | 'val2' | 'val3' | 'val4';\n` +
          `export type enum2 = 'val5' | 'val6' | 'val7' | 'val8';\n`
      );
    });
  });

  describe('generateEnumType', () => {
    it('empty object', () => {
      const enumType = Typescript.generateEnumType({}, options);
      expect(enumType).toBe('');
    });

    it('with enumerations', () => {
      const enumType = Typescript.generateEnumType(
        {
          enum1: ['val1', 'val2', 'val3', 'val4'],
          enum2: ['val5', 'val6', 'val7', 'val8'],
        },
        options
      );

      expect(enumType).toBe(
        `export type enum1 = 'val1' | 'val2' | 'val3' | 'val4';\n` +
          `export type enum2 = 'val5' | 'val6' | 'val7' | 'val8';\n`
      );
    });
  });
  describe('generateTableTypes', () => {
    it('empty table definition object', () => {
      const tableTypes = Typescript.generateTableTypes('tableName', {}, options);
      expect(tableTypes).toMatchSnapshot();
    });

    it('with table definitions', () => {
      const tableTypes = Typescript.generateTableTypes(
        'tableName',
        {
          col1: { udtName: 'name1', nullable: false, tsType: 'string' },
          col2: { udtName: 'name2', nullable: false, tsType: 'number' },
        },
        options
      );
      expect(tableTypes).toMatchSnapshot();
    });

    it('with nullable column definitions', () => {
      const tableTypes = Typescript.generateTableTypes(
        'tableName',
        {
          col1: { udtName: 'name1', nullable: true, tsType: 'string' },
          col2: { udtName: 'name2', nullable: true, tsType: 'number' },
        },
        options
      );

      expect(tableTypes).toMatchSnapshot();
    });
  });
});
