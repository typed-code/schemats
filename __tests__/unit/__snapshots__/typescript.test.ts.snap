// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typescript generateTableInterface empty table definition object 1`] = `
"
        export interface tableName {
        
        }
    "
`;

exports[`Typescript generateTableInterface table name is reserved 1`] = `
"
        export interface package_ {
        
        }
    "
`;

exports[`Typescript generateTableInterface table with columns 1`] = `
"
        export interface tableName {
        col1: tableNameFields.col1;
col2: tableNameFields.col2;

        }
    "
`;

exports[`Typescript generateTableInterface table with reserved columns 1`] = `
"
        export interface tableName {
        string: tableNameFields.string_;
number: tableNameFields.number_;
package: tableNameFields.package_;

        }
    "
`;

exports[`Typescript generateTableTypes empty table definition object 1`] = `
"
        export namespace tableNameFields {
        
        }
    "
`;

exports[`Typescript generateTableTypes with nullable column definitions 1`] = `
"
        export namespace tableNameFields {
        export type col1 = string| null;
export type col2 = number| null;

        }
    "
`;

exports[`Typescript generateTableTypes with table definitions 1`] = `
"
        export namespace tableNameFields {
        export type col1 = string;
export type col2 = number;

        }
    "
`;
